lib/internal/webstreams/adapters.js

```diff
@@ -0,0 +1,746 @@
+    closed = true;
+    streamReadable.off('close', onClose);
+    streamReadable.off('data', onData);
+    streamReadable.off('end', onEnd);
+    streamReadable.off('error', onError);
+  }
+
+  function onError(error) {
+    controller.error(error);
+  }
+
+  function onData(chunk) {
+    // Copy the Buffer to detach it from the pool.
+    if (Buffer.isBuffer(chunk))
+      chunk = new Uint8Array(chunk);
+    controller.enqueue(chunk);
+  }
+
+  function onEnd() {
+    controller.close();
+  }
+
+  streamReadable.once('close', onClose);
+  streamReadable.once('end', onEnd);
+  streamReadable.once('error', onError);
+  streamReadable.on('data', onData);
+  streamReadable.pause();
+
+  return new ReadableStream({
+    type,
+    autoAllocateChunkSize,
+
+    start(c) { controller = c; },
+
+    pull() { streamReadable.resume(); },
+
+    cancel(reason) {
+      if (!closed)
+        streamReadable.destroy(reason);
+    },
+  }, strategy);
+}
+
+function newReadableWritablePairFromDuplex(duplex) {
+  if (typeof duplex?._writableState !== 'object' ||
+      typeof duplex?._readableState !== 'object') {
+    throw new ERR_INVALID_ARG_TYPE('duplex', 'stream.Duplex', duplex);
+  }
+
+  const writable = newWritableStreamFromStreamBase(duplex);
+  const readable = newReadableStreamFromStreamBase(duplex);
+
+  return {
+    writable,
+    readable,
+  };
+}
+
+function newStreamWritableFromWritableStream(writableStream, options = {}) {
+  if (!isWritableStream(writableStream)) {
+    throw new ERR_INVALID_ARG_TYPE(
+      'writableStream',
+      'WritableStream',
+      writableStream);
+  }
+
+  validateObject(options, 'options');
+  const {
+    objectMode = false,
+    highWaterMark,
+    signal,
+  } = options;
+
+  validateBoolean(objectMode, 'options.objectMode');
+  if (signal !== undefined)
+    validateAbortSignal(signal, 'options.signal');
+
+  const writer = writableStream.getWriter();
+  let closed = false;
+
+  const writable = new Writable({
+    highWaterMark: highWaterMark || writer.desiredSize,
+    objectMode,
+
+    writev(chunks, callback) {
+      if (closed) {
+        callback(new ERR_INVALID_STATE('The WritableStream is closed'));
+        return;
+      }
+
+      function done(error) {
+        try {
+          callback(error);
+        } catch (error) {
+          process.nextTick(() => writable.destroy(error));
+        }
+      }
+
+      PromisePrototypeThen(
+        writer.ready,
+        () => {
+          return PromisePrototypeThen(
+            PromiseAll(
+              ArrayPrototypeMap(
+                chunks,
+                (chunk) => writer.write(chunk))),
+            done,
+            done);
+        },
+        done);
+    },
+
+    write(chunk, encoding, callback) {
+      if (closed) {
+        callback(new ERR_INVALID_STATE('The WritableStream is closed'));
+        return;
+      }
+      if (typeof chunk === 'string')
+        chunk = new Uint8Array(Buffer.from(chunk, encoding));
```

Comment:
This should only happen if `objectMode == false`.

That said, is this even necessary? If I'm reading [the documentation](https://nodejs.org/api/stream.html#stream_writable_write_chunk_encoding_callback_1) correctly, `chunk` can only be a string if `decodeStrings == false` or if `objectMode == true`. Since we never set `decodeStrings` to `false`, `chunk` will always be a `Buffer` if `objectMode == false`.